### PR TITLE
[libclc] Update build system to reflect LLVM upstream changes

### PR DESCRIPTION
**Pull request by frasercrmck**
This brings the libclc build system more in line with LLVM upstream, which was recently changed significantly. It throws out a fair bit of dead CMake code too.

Upstream, libclc was changed away from using custom languages to build and towards custom commands. This makes in-tree builds a lot simpler and less hacky, as well as a few other benefits.

Unfortunately, due to peculiarities involving the `-c` and `-x cl` flags, and the combination of the two which enable/disable different optimizations, I was seeing significant changes to the builtin libraries when using the same flags as upstream. As such, I've had to diverge from upstream and not use `-x cl` when compiling builtin files. This should keep the builtin libraries identical to before this change, but it's not ideal to be compiling ostensibly OpenCL files as if they were another language.

The intermediate bytecode files and builtins have moved a bit, but the final libraries (both remangled and non-remangled) are in the same place.

In the process, a bug was fixed that was misleading users into thinking a custom `libclc-remangler` tool was used when the `LIBCLC_CUSTOM_LLVM_TOOLS_BINARY_DIR` was set, when in fact the locally built one was always being used.

### CODE REVIEW

**Comment by frasercrmck** [comment]
CC @jsji

**Comment by jsji** [comment]
Looks like check-libclc is failing due to missing spirv headers. "#include <spirv/spirv_types.h>"

**Comment by ldrumm** [comment]
image
Good job github; I clearly have no interest in reviewing "large" changes ;)

**Comment by frasercrmck** [comment]
> Looks like check-libclc is failing due to missing spirv headers. "#include <spirv/spirv_types.h>"

Yep, I think I found the problem. :crossed_fingers:

File: libclc/generic/libspirv/SOURCES [hunk 1]
```diff
@@ -44,13 +44,10 @@ core/shared/clc_max.cl
 core/shared/clc_min.cl
 geometric/cross.cl
 geometric/distance.cl
-geometric/distance.inc
 geometric/dot.cl
 geometric/fast_distance.cl
-geometric/fast_distance.inc
```
Comment (ldrumm) [inline libclc/generic/libspirv/SOURCES:50]: Why are these removed?

File: libclc/utils/libclc-remangler/CMakeLists.txt [hunk 2]
```diff
@@ -24,3 +24,9 @@ clang_target_link_libraries(libclc-remangler
   clangSerialization
   LLVMOption
   )
+
+# If we've not already imported a libclc-remangler tool from an external build,
+# set it up now.
```
Comment (ldrumm) [inline libclc/utils/libclc-remangler/CMakeLists.txt:29]: Is this a new feature? that we can now use an existing remangler?

File: libclc/CMakeLists.txt [hunk 3]
```diff
@@ -1,47 +1,110 @@
...
+  ptx-nvidiacl/libspirv/SOURCES;
   r600/lib/SOURCES;
+  r600/libspirv/SOURCES;
   spirv/lib/SOURCES;
   spirv64/lib/SOURCES
+  x86_64-unknown-linux/libspirv/SOURCES
 )
 
-include( AddLibclc )
+set( LIBCLC_MIN_LLVM 3.9.0 )
```
Comment (ldrumm) [inline libclc/CMakeLists.txt:38]: This is new. Have you tested 3.9?

File: libclc/CMakeLists.txt [hunk 4]
```diff
@@ -1,47 +1,110 @@
...
+
+set( LIBCLC_TARGETS_TO_BUILD "all"
+    CACHE STRING "Semicolon-separated list of targets to build, or 'all'." )
```
Comment (ldrumm) [inline libclc/CMakeLists.txt:41]: Might want to mention libclc in the help message, because that wording matches LLVM_TARGETS_TO_BUILD and LLVM_ENABLE_PROJECTS somewhat, and that might be a bit confusing on the command line

File: libclc/CMakeLists.txt [hunk 5]
```diff
@@ -1,47 +1,110 @@
...
+  include(AddLLVM)
+
+  message( STATUS "libclc LLVM version: ${LLVM_PACKAGE_VERSION}" )
+
+  if( LLVM_PACKAGE_VERSION VERSION_LESS LIBCLC_MIN_LLVM )
+    message( FATAL_ERROR "libclc needs at least LLVM ${LIBCLC_MIN_LLVM}" )
+  endif()
+
+  # Import required tools as targets
+  if ( NOT EXISTS LIBCLC_CUSTOM_LLVM_TOOLS_BINARY_DIR )
```
Comment (ldrumm) [inline libclc/CMakeLists.txt:59]: nice way of handling these binaries!

File: libclc/CMakeLists.txt [hunk 6]
```diff
@@ -1,47 +1,110 @@
...
+  set( LLVM_PACKAGE_VERSION ${LLVM_VERSION} )
+
+  # Note that we check this later (for both build types) but we can provide a
+  # more useful error message when built in-tree. We assume that LLVM tools are
+  # always available so don't warn here.
+  if( NOT clang IN_LIST LLVM_ENABLE_PROJECTS )
+    message(FATAL_ERROR "Clang is not enabled, but is required to build libclc in-tree")
+  endif()
+
+  if ( NOT EXISTS LIBCLC_CUSTOM_LLVM_TOOLS_BINARY_DIR )
```
Comment (ldrumm) [inline libclc/CMakeLists.txt:79]: There's basically an `else` clause below, but it's a standalone if clause gaurded by the opposite condition. Perhaps you could be a diamond geezer, and turn this into an if-else?

File: libclc/CMakeLists.txt [hunk 7]
```diff
@@ -1,47 +1,110 @@
...
+    " ensure the tools are up to date." )
+  foreach( tool clang llvm-as llvm-link opt llvm-spirv libclc-remangler )
+    find_program( LLVM_TOOL_${tool} ${tool}
+      PATHS ${LIBCLC_CUSTOM_LLVM_TOOLS_BINARY_DIR} NO_DEFAULT_PATH )
+  endforeach()
+
+  # If we've requested a custom binary directory, there are some otherwise
+  # optional tools which we want to ensure are present.
+  if( NOT TARGET libclc::llvm-spirv OR NOT TARGET libclc::libclc-remangler )
+    message( FATAL_ERROR "libclc toolchain incomplete!" )
```
Comment (ldrumm) [inline libclc/CMakeLists.txt:98]: please include what's missing in the error message

File: libclc/CMakeLists.txt [hunk 8]
```diff
@@ -52,170 +115,51 @@ set( LIBCLC_TARGETS_ALL
...
   list( APPEND LIBCLC_TARGETS_ALL amdgcn-mesa-mesa3d )
 endif()
-# $&$ end delete
 
-set( LIBCLC_TARGETS_TO_BUILD "all"
-    CACHE STRING "Semicolon-separated list of targets to build, or 'all'." )
-
-option( ENABLE_RUNTIME_SUBNORMAL "Enable runtime linking of subnormal support."
-OFF )
+# spirv-mesa3d and spirv64-mesa3d targets can only be built with the (optional)
```
Comment (ldrumm) [inline libclc/CMakeLists.txt:130]: above we throw a fatal error if we can't find these tools. Are we sure they're optional?

**Review by ldrumm — APPROVED** [review]
This is nice work! A couple of things I'm unsure about, but don't let my questions stop this being merged.

**Comment by ldrumm** [comment]
s/Upstream, libclc was changed away from using custom languages to built/Upstream, libclc was changed away from using custom languages to build/

?

File: libclc/CMakeLists.txt (hunk 3 above)
Comment (frasercrmck) [reply libclc/CMakeLists.txt:38]: It's new to us but it shouldn't be - we've deleted it. [Upstream](https://github.com/llvm/llvm-project/blob/main/libclc/CMakeLists.txt#L32), `libclc` is (nominally) intended to work on all of these LLVMs. Obviously that's not tested so it's theoretical as far as I can tell. However, if we want to eventually merge back upstream I don't think we should be dropping things like this. That's why I "added" it back in.

File: libclc/CMakeLists.txt (hunk 4 above)
Comment (frasercrmck) [reply libclc/CMakeLists.txt:41]: Yeah that's probably a good idea. I haven't changed it in this PR, though. I think I'll open a PR with upstream and that'll filter back down to us. Does that sound okay?

File: libclc/CMakeLists.txt (hunk 7 above)
Comment (frasercrmck) [reply libclc/CMakeLists.txt:98]: Good idea. Unfortunately it's like this [upstream](https://github.com/llvm/llvm-project/blob/main/libclc/CMakeLists.txt#L78). I'll open a PR there and it'll filter down to us in time. Sound alright?

Comment (frasercrmck) [reply libclc/CMakeLists.txt:98]: Oh but these tools in particular are specific to us. So yeah maybe I can include the missing tools here, not below, and open up that PR.

File: libclc/CMakeLists.txt (hunk 8 above)
Comment (frasercrmck) [reply libclc/CMakeLists.txt:130]: Upstream the tool definitely is optional. For us it's not optional if you use the `LIBCLC_CUSTOM_LLVM_TOOLS_BINARY_DIR` option which upstream doesn't have.

So we handle the requirements on that tool slightly differently, but again we've been deleting or not integrating upstream code (handling these targets and others like [clspv/clspv64](https://github.com/llvm/llvm-project/blob/main/libclc/CMakeLists.txt#L22-L23) - [for example](https://github.com/llvm/llvm-project/blob/main/libclc/CMakeLists.txt#L254-L256)) that we'll need to be able to accommodate if we want to merge our changes back to LLVM.

File: libclc/generic/libspirv/SOURCES (hunk 1 above)
Comment (frasercrmck) [reply libclc/generic/libspirv/SOURCES:50]: We don't actually want to compile these `.inc` files to their own bytecode files, which is the primary purpose of these `SOURCES` files. Nor indeed can we - they're not complete CL files and are usually included from other files.

With the previous approach I think these were just silently ignored somehow.

I suspect someone added them for dependency purposes, but we don't need to do that anymore since we have `-MD` and pals to do the job for us.

File: libclc/CMakeLists.txt (hunk 4 above)
Comment (ldrumm) [reply libclc/CMakeLists.txt:41]: ack

File: libclc/generic/libspirv/SOURCES (hunk 1 above)
Comment (ldrumm) [reply libclc/generic/libspirv/SOURCES:50]: I figured that some include machinery was involved. Thanks for the explanation

File: libclc/CMakeLists.txt (hunk 6 above)
Comment (ldrumm) [reply libclc/CMakeLists.txt:79]: fair

File: libclc/CMakeLists.txt (hunk 8 above)
Comment (ldrumm) [reply libclc/CMakeLists.txt:130]: ack

File: libclc/utils/libclc-remangler/CMakeLists.txt (hunk 2 above)
Comment (frasercrmck) [reply libclc/utils/libclc-remangler/CMakeLists.txt:29]: I don't believe it's a new feature, it should currently be handled by the `LIBCLC_CUSTOM_LLVM_TOOLS_BINARY_DIR` option importing `libclc-remangler` from another location. However since I realise I've never tried that option myself, and I don't know how it works (shouldn't there be a clash on the `libclc-remangler` target?) I've kicked off a build locally and will get back to you once I work out what's going on.

**Comment by frasercrmck** [comment]
> s/Upstream, libclc was changed away from using custom languages to built/Upstream, libclc was changed away from using custom languages to build/
> 
> ?

Thanks, clippy!

File: libclc/utils/libclc-remangler/CMakeLists.txt (hunk 2 above)
Comment (frasercrmck) [reply libclc/utils/libclc-remangler/CMakeLists.txt:29]: Aye unfortunately it seems that there's currently a bug where `LIBCLC_CUSTOM_LLVM_TOOLS_BINARY_DIR` *says* that the `libclc-remangler` it taken from that path, but in fact it's always using the built one. It's because we set the `LIBCLC_REMANGLER` variable, but then use the `libclc-remangler` target to build.

I think we can consider this change a bug fix. I'll update the PR description.

File: libclc/CMakeLists.txt (hunk 7 above)
Comment (frasercrmck) [reply libclc/CMakeLists.txt:98]: Done!
